File: YouDontMakeFriendsWithDotNet/NoRightAnswersToWrongQuestions/AntlrInterpretor/Logic/QlErrorListener.cs
using Antlr4.Runtime;
using QuestionaireDomain.Entities.API;

namespace AntlrInterpretor.Logic
{
    internal class QlErrorListener : BaseErrorListener, IAntlrErrorListener<int>
    {
        public override void SyntaxError(
            IRecognizer recognizer, 
            IToken offendingSymbol, 
            int line,
            int charPositionInLine, 
            string msg, 
            RecognitionException e)
        {
            const string message = @"Parse failed. See inner exception for details.";
            var detailsMessage = $@"'{offendingSymbol.Text}' was not recognized at line {line}, position {charPositionInLine}, giving the following error: {msg} ";
            throw new QlParserException(message, e)
            {
                ParseErrorDetails = detailsMessage
            };
        }
        
        public void SyntaxError(
            IRecognizer recognizer, 
            int offendingSymbol, 
            int line, 
            int charPositionInLine, 
            string msg,
            RecognitionException e)
        {
            const string message = @"Parse failed. See inner exception for details.";
            var detailsMessage = $@"Lexing of {recognizer.InputStream} failed at line {line}, position {charPositionInLine}, giving the following error: {msg} ";
            throw new QlParserException(message, e)
            {
                ParseErrorDetails = detailsMessage
            };
        }
    }
}

File: durmitor/src/main/java/ql/listeners/QLErrorListener.java
package ql.listeners;

import java.util.ArrayList;
import java.util.List;

import org.antlr.v4.runtime.BaseErrorListener;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Recognizer;

import ql.exceptions.ANTLRError;
import ql.exceptions.QLException;

public class QLErrorListener extends BaseErrorListener {

    private List<QLException> errors;

    public QLErrorListener() {
        this.errors = new ArrayList<QLException>();
    }
    
    public QLErrorListener(List<QLException> errors) {
        this.errors = errors;
    }

    @Override
    public void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e) {
        errors.add(new ANTLRError(msg, line, charPositionInLine));
    }

    public boolean hasErrors() {
        return !errors.isEmpty();
    }

    public void printErrors() {
        for (QLException e : errors) System.out.println(e);
    }
}
